addons/account/models/account_account.py

```diff
@@ -134,28 +135,53 @@ def _check_account_type_unique_current_year_earning(self):
+            query.add_join('LEFT JOIN', alias=company_dependent_field_alias, table='ir_property', condition=SQL(
+                    """
+                        %(ir_property_fields_id)s = %(field_id)s
+                        AND %(ir_property_company_id)s = %(root_company_id)s
+                        AND %(ir_property_res_id)s = 'account.account,' || %(account_id)s::text
+                    """,
+                    ir_property_fields_id=self.env['ir.property']._field_to_sql(company_dependent_field_alias, 'fields_id'),
+                    field_id=self.env['ir.model.fields']._get(self._name, field.name).id,
+                    ir_property_company_id=self.env['ir.property']._field_to_sql(company_dependent_field_alias, 'company_id'),
+                    root_company_id=self.env.company.root_id.id,
+                    ir_property_res_id=self.env['ir.property']._field_to_sql(company_dependent_field_alias, 'res_id'),
+                    account_id=SQL.identifier(alias, 'id'),
+                )
+            )
+
+            # This is the field on ir.property that will contain the value of 'code'.
+            value_field = self.env['ir.property']._fields[TYPE2FIELD[field.type]]
+
+            return self.env['ir.property']._field_to_sql(company_dependent_field_alias, value_field.name)
+
         return super()._field_to_sql(alias, fname, query, flush)
 
-    @api.constrains('company_id', 'code')
+    @api.constrains('company_ids')
+    def _constrain_at_least_one_company(self):
+        if accounts_without_company := self.filtered(lambda a: not a.sudo().company_ids):
+            raise ValidationError(
+                _("The following accounts must be assigned to at least one company:")
+                + "\n" + "\n".join(f"- {account.display_name}" for account in accounts_without_company)
+            )
+
+    @api.constrains('company_ids', 'code')
     def _constrains_code(self):
-        # check for duplicates in each root company
-        by_root_company = self.grouped(lambda record: record.company_id.root_id)
-        for root_company, records in by_root_company.items():
-            by_code = records.grouped('code')
-            if len(by_code) < len(records):
-                # retrieve duplicates within self
-                duplicates = next(recs for recs in by_code.values() if len(recs) > 1)
-            else:
-                # search for duplicates of self in database
-                duplicates = self.search([
-                    ('company_id', 'child_of', root_company.id),
-                    ('code', 'in', list(by_code)),
-                    ('id', 'not in', records.ids),
-                ])
+        """ Ensure that for each company, all non-False account codes are unique. """
+        duplicates = {}
+        for company in self.env.companies:
+            accounts_with_code = self.sudo().with_company(company).filtered(lambda a: a.code)
```

Comment:
```suggestion
        for company in self.env.companies.root_id.sudo():
            accounts_with_code = self.with_company(company).filtered(lambda a: a.code)
```

Reply:
I agree with checking only on the root companies, however, we do need to call `sudo()` on the account rather than on the company...